staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/implementation.go

```diff
@@ -0,0 +1,273 @@
+		return a.object.Object, true
+	case OldObjectVarName:
+		return a.oldObject.Object, a.hasOld
+	case ParamsVarName:
+		return a.params.Object, true
+	case RequestVarName:
+		return a.request.Object, true
+	default:
+		return nil, false
+	}
+}
+
+// Parent returns the parent of the current activation, may be nil.
+// If non-nil, the parent will be searched during resolve calls.
+func (a *validationActivation) Parent() interpreter.Activation {
+	return nil
+}
+
+// Compile compiles the cel expression defined in ValidatingAdmissionPolicy
+func (c *CELValidatorCompiler) Compile(p *v1alpha1.ValidatingAdmissionPolicy) (Validator, error) {
+	if len(p.Spec.Validations) == 0 {
+		return nil, nil
+	}
+	hasParam := false
+	if p.Spec.ParamKind != nil {
+		hasParam = true
+	}
+	compilationResults := make([]CompilationResult, len(p.Spec.Validations))
+	for i, validation := range p.Spec.Validations {
+		compilationResults[i] = CompileValidatingPolicyExpression(validation.Expression, hasParam)
+	}
+	return &CELValidator{policy: p, compilationResults: compilationResults}, nil
+}
+
+// CELValidator implements the Validator interface
+type CELValidator struct {
+	policy             *v1alpha1.ValidatingAdmissionPolicy
+	compilationResults []CompilationResult
+}
+
+func convertObjectToUnstructured(obj interface{}) *unstructured.Unstructured {
+	var unstructuredObj unstructured.Unstructured
+	jsonObj, _ := json.Marshal(obj)
+	_ = json.Unmarshal(jsonObj, &unstructuredObj.Object)
+
+	return &unstructuredObj
+}
+
+// Validate validates all cel expressions in Validator and returns PolicyDecisions and error.
+func (v *CELValidator) Validate(a admission.Attributes, o admission.ObjectInterfaces, params *unstructured.Unstructured) ([]PolicyDecision, error) {
+	decisions := make([]PolicyDecision, len(v.policy.Spec.Validations))
+	var oldObjectVal *unstructured.Unstructured
+	if a.GetOldObject() != nil {
+		oldObjectVal = convertObjectToUnstructured(a.GetOldObject())
+	}
+
+	var objectVal *unstructured.Unstructured
+	if a.GetObject() != nil {
+		objectVal = convertObjectToUnstructured(a.GetObject())
+	}
+	va := &validationActivation{
+		object:    objectVal,
+		oldObject: oldObjectVal,
+		params:    params,
+		request:   convertObjectToUnstructured(createAdmissionRequest(a)),
+	}
+
+	for i, compilationResult := range v.compilationResults {
+		var policyDecision = &decisions[i]
+
+		if compilationResult.Error != nil {
+			policyDecision.Kind = Deny
+			policyDecision.Message = fmt.Sprintf("compilation error: %v", compilationResult.Error)
+			continue
+		}
+		if compilationResult.Program == nil {
+			continue
+		}
+		evalResult, _, err := compilationResult.Program.Eval(va)
+		if err != nil {
+			f := *v.policy.Spec.FailurePolicy
+			if f == v1alpha1.Ignore {
+				policyDecision.Kind = Admit
+			} else {
+				policyDecision.Kind = Deny
+			}
+			policyDecision.Message = fmt.Sprintf("failed expression: %v, with evaluation error: %v", v.policy.Spec.Validations[i].Expression, err)
```

Comment:
Message should only be set for Deny case?

Reply:
Now i have updated to also use message for logging when FailurePolicy is Ignore so we will not lost the err info